### DISCUSSION TITLE
Countries and cities are not resolved (kubernetes with nginx-ingress)

### DISCUSSION THREAD

**Discussion by miwurster** [plausible/analytics#3207] [Self-Hosted Support] [answered]
Hello, 
we have deployed Plausible into our Kubernetes cluster while using `nginx-ingress` as proxy. We already tried to configure the ingress to Plausible with `proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;`. 

However, in Plausible we do not see any resolved country or city.

Has anyone of you a tip/hint how to debug this or even resolve the issue?

Thanks in advance,
Michael

**Comment by ruslandoga** [comment]
👋 @miwurster 

Something like https://github.com/plausible/analytics/discussions/2767#discussioncomment-5367417 can be used to debug the headers Plausible receives.

**Comment by miwurster** [comment]
Awesome tip! Thank you very much. 

I can see that Plausible receives the `x-forwarded-to` header. However, the value is, for example:
```
{"x-forwarded-for", "10.80.3.1, 10.80.3.1"}
```

That's weird, it is the ip address range of the Kubernetes cluster itself. Anyone has a clue what is going on?

**Reply by ruslandoga** [reply, marked as answer]
<details>
<summary>Here's what GPT says...</summary>

It appears that the X-Forwarded-For header is showing the internal IP address of the Kubernetes cluster. This is likely because the request is going through multiple layers of proxies (in this case, Nginx and Kubernetes services), and each proxy appends the IP address of the previous client (which can be a user or another proxy) to the X-Forwarded-For header.

In your case, it appears that Nginx is appending the IP address of the Kubernetes node (or pod) that sent the request, rather than the original client IP. This can happen if you haven't configured Nginx and/or Kubernetes correctly to forward the original client IP.

To resolve this issue, you should ensure the following:

1. Make sure your Nginx ingress controller is configured to use the `use-forwarded-headers` or `compute-full-forwarded-for` setting. These settings tell Nginx to use the original client IP from the X-Forwarded-For header, rather than the IP of the Kubernetes node.

2. If you're using a Kubernetes service to proxy traffic to Nginx, you should configure it to use `externalTrafficPolicy: Local`. This tells Kubernetes to preserve the original client IP when forwarding requests to the service. 

Here's how you can set it in your Service configuration:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: LoadBalancer
  externalTrafficPolicy: Local
  ...
```

3. Depending on your cloud provider or networking solution, you may need additional configuration. For example, in Google Kubernetes Engine, you might need to enable the `X-Forwarded-For` trust settings.

Please apply these settings and let me know how it goes! If you still have the issue, I'll be here to assist further.

</details>

**Reply by miwurster** [reply]
Haha nice, I had the same idea in the meantime. Got a similar answer. Atm I'm experimenting with `use-forwarded-headers` and `compute-full-forwarded-for`, but with no luck. Looking into `externalTrafficPolicy` now. Keep you posted.

Thanks for your input. Really appreciate it!!

**Reply by miwurster** [reply]
The option `externalTrafficPolicy` did the trick! It works now!
